### ISSUE TITLE
Add `patch` in addition to `set`

### ISSUE THREAD

**Issue by Vadorequest** [zydalabs/swr-internal-state#2]
When managing concurrent writes, using `set` can be dangerous, because we might lose data if 2 concurrent calls are performed at the same time.

Would you consider implementing a "patch" function as 3rd argument? Such as:

`const [userInfo, setUserInfo, patchUserInfo] = useUserInfo();`

I think it makes sense (especially for global object) to provide an alternative. Everybody provides a `setState` function, but that's not concurrent-friendly.

**Comment by AmrSaber** [comment]
Can you elaborate more on how that function should behave in normal case and in case of concurrent updates?

**Comment by Vadorequest** [comment]
I'm not sure. The core issue I have with `setState` right now is that it requires the whole state.

When performing `setState` with concurrency, it happens sometimes that part of the state is lost, because the latest `setState` wins when the state used was a copy which got out of date. (the copy isn't updated when the state is changed)

I'm talking about issues I'm encountering when dealing with concurrent calls that are 1-5ms close to each other, it's a particular use case/edge case.

The idea would be not to set the whole state but only to mutate the data that are meant to be changed (patching, basically).

With Redux, there is no setState, it's a reducer that always uses the latest state and only applies a patch on top of it.

With Recoil/React.useState they use a setState which isn't concurrent-friendly.

Does that help?

**Comment by AmrSaber** [comment]
Yes, this clarifies things. I have an idea in mind, but I am not sure if it would do the job or not.

What I think about the implementation is that we would have a function with the signature
```ts
patchState(path: string, value: any): void
```
That `path` would be similar to what lodash's [_.set](https://lodash.com/docs/4.17.15#set) expects. Now the problem would be with the function's logic. The data is saved in SWR as a bulk state, and any update has to be on that whole state (think of it as a key-value store). So the function would just update the state as you would do in your code, it would only ensure that the update is made on the latest state that is saved.

But this would not be much different from what would do in your code, something like
```ts
const [myState] = useSomeState();

// Some logic..
myState.x = getSomeValue();

// Update x in the state
const [latestState, setState] = useSomeState();
setState({ ...state, x: myState.x });
```
This is a little different from how you would handle normal react's state though, as here you can make several calls to the same state hook and it would give you the same value.

I think just following the above approach would solve your problem (partially?). I do see some benefits of adding this to the package, but I am not sure if we should, what do you think?

**Comment by Vadorequest** [comment]
Yeah, I see. It's hard for me to tell if it's outside the scope of the lib, though.

It really depends upon whether handling concurrent writes is something that must be supported.
For the moment, what I did is having a "mutations queue" where all patches are added, and are applied in batch. So, I eventually call `setState`, but I'm always using the latest value of the state and all changes are "piped/queued" to make sure they're executed in order.

This also affects the whole app, because I can't mix both "patch" and "set" concept, as any use of a "set" would override the "patch" queue. So, maybe that'd actually make more sense to provide a completely different API for this, to avoid providing both set and patch mutators/accessors, but only one of them.

**Comment by AmrSaber** [comment]
I like the idea of the queue, maybe with a `setTimeOut` that gets sets with a small interval whenever the queue is not empty, and that gets cleared when the queue is empty.

I don't see why we wouldn't support such a feature. Maybe it can also be used by a separate interface as you said and it can use the existing `useGlobalState` hook.

Would you like to make a PR with this new feature?

**Comment by Vadorequest** [comment]
I used `useEffect` to run the execution of the queue and cleaning.

https://github.com/Vadorequest/rwa-faunadb-reaflow-nextjs-magic/blob/main/src/utils/canvasDatasetMutationsQueue.ts
https://github.com/Vadorequest/rwa-faunadb-reaflow-nextjs-magic/blob/main/src/components/editor/CanvasContainer.tsx#L123

Unfortunately, I won't have time to implement such feature.

**Comment by AmrSaber** [comment]
I checked your implementation, it's very elegant indeed. But I think we would require a more compact approach for a package, it's not a good thing to have a feature spread on several parts.

I think such a patch can be implemented with a queue that is saved in a global state on its own. and the logic for enqueueing elements should trigger a `setTimeout` or so to start executing the mutations from that queue. Then when the queue is empty the `setTimeout` can be cleared. I don't think we would need to use `useEffect` as we affect the state directly and not any existing components.

All of this is still theoretical, but I think we can give it a shot, it would be fun to implement :)

**Comment by Vadorequest** [comment]
Ahah, I'm glad you find it fun, I hope other users will find it useful too! 😉
